File: dp_tabulation/0552_hard/student_attendance.py

def count_possibilities(record_len: int) -> int:  # LeetCode Q.552.
    modulo = 10 ** 9 + 7

    # Possible suffixes: "a", "p", "l", "ll".
    one_absence_table = {"a": 1, "p": 0, "l": 0, "ll": 0}  # Base case: record len = 1.

    # Possible suffixes: "p", "l", "ll".
    no_absence_table = {"p": 1, "l": 1, "ll": 0}  # Base case: record len = 1.

    # Help each day's tables update values.
    new_one_absence_table, new_no_absence_table = dict(), dict()

    for _ in range(2, record_len + 1):
        # Update counts of those w/ 1 absence and suffix "a".
        new_one_absence_table["a"] = sum(no_absence_table.values()) % modulo

        # Update counts of those w/ 1 absence and suffix "p".
        new_one_absence_table["p"] = sum(one_absence_table.values()) % modulo

        # Update counts of those w/ 1 absence and suffix "l".
        new_one_absence_table["l"] = one_absence_table["a"] + one_absence_table["p"]

        # Update counts of those w/ 1 absence and suffix "ll".
        new_one_absence_table["ll"] = one_absence_table["l"]

        # Update counts of those w/o absence and has suffix "l" or "ll".
        new_no_absence_table["l"] = no_absence_table["p"]
        new_no_absence_table["ll"] = no_absence_table["l"]

        # Update counts of those w/o absence and has suffix "p".
        new_no_absence_table["p"] = sum(no_absence_table.values()) % modulo

        one_absence_table.update(new_one_absence_table)
        no_absence_table.update(new_no_absence_table)

        new_one_absence_table.clear()  # Reset for next day's usage.
        new_no_absence_table.clear()

    valid_combinations = sum(one_absence_table.values()) + sum(no_absence_table.values())
    valid_combinations %= modulo
    return valid_combinations

File: docs/dp_tabulation/0552_hard/student_attendance.cpp
#include <vector>
#include <cmath>
using namespace std;

int countPossibilities(int recordLen) // LeetCode Q.552.
{
    long long modulo = pow(10, 9) + 7; // Long long prevents overflow.

    // Possible suffixes: "a", "p", "l", "ll".
    long long oneAbsenceSuffixA = 1;
    long long oneAbsenceSuffixP = 0, oneAbsenceSuffixL = 0, oneAbsenceSuffixLL = 0;

    // Possible suffixes: "p", "l", "ll".
    long long noAbsenceSuffixP = 1, noAbsenceSuffixL = 1;
    long long noAbsenceSuffixLL = 0;

    // Helper variables: help each day's branch counts update values.
    long long newOneAbsenceSuffixA = 0, newOneAbsenceSuffixP = 0;
    long long newOneAbsenceSuffixL = 0, newOneAbsenceSuffixLL = 0;

    long long newNoAbsenceSuffixP = 0;
    long long newNoAbsenceSuffixL = 0, newNoAbsenceSuffixLL = 0;

    for (int day = 2; day <= recordLen; day++)
    {
        // Update counts of those w/ one absence suffix "a".
        newOneAbsenceSuffixA = (noAbsenceSuffixL + noAbsenceSuffixLL + noAbsenceSuffixP) % modulo;

        // Update counts of those w/ one absence suffix "p".
        newOneAbsenceSuffixP = (oneAbsenceSuffixL + oneAbsenceSuffixLL + oneAbsenceSuffixA + oneAbsenceSuffixP) % modulo;

        // Update counts of those w/ 1 absence and suffix "l".
        newOneAbsenceSuffixL = (oneAbsenceSuffixA + oneAbsenceSuffixP) % modulo;

        // Update counts of those w/ 1 absence and suffix "ll".
        newOneAbsenceSuffixLL = oneAbsenceSuffixL;

        // Update counts of those w/o absence and has suffix "l" or "ll".
        newNoAbsenceSuffixL = noAbsenceSuffixP;
        newNoAbsenceSuffixLL = noAbsenceSuffixL;

        // Update counts of those w/o absence and has suffix "p".
        newNoAbsenceSuffixP = (noAbsenceSuffixP + noAbsenceSuffixL + noAbsenceSuffixLL) % modulo;

        // Transit updated counts.
        oneAbsenceSuffixA = newOneAbsenceSuffixA;
        oneAbsenceSuffixP = newOneAbsenceSuffixP;
        oneAbsenceSuffixL = newOneAbsenceSuffixL;
        oneAbsenceSuffixLL = newOneAbsenceSuffixLL;

        noAbsenceSuffixP = newNoAbsenceSuffixP;
        noAbsenceSuffixL = newNoAbsenceSuffixL;
        noAbsenceSuffixLL = newNoAbsenceSuffixLL;

        // Reset helper variables for next day's usage.
        newOneAbsenceSuffixA = 0, newOneAbsenceSuffixP = 0;
        newOneAbsenceSuffixL = 0, newOneAbsenceSuffixLL = 0;
        newNoAbsenceSuffixP = 0, newNoAbsenceSuffixL = 0, newNoAbsenceSuffixLL = 0;
    }

    long long validCombinations = (oneAbsenceSuffixA + oneAbsenceSuffixP + oneAbsenceSuffixL + oneAbsenceSuffixLL) % modulo;

    validCombinations += (noAbsenceSuffixP + noAbsenceSuffixL + noAbsenceSuffixLL) % modulo;

    return validCombinations % modulo;
}
